src/types/from_node.rs

```diff
@@ -0,0 +1,357 @@
+        });
+        Ok(result)
+    }
+}
+
+impl TryFrom<TextureLayout> for builtin_transformations::TextureLayout {
+    type Error = anyhow::Error;
+
+    fn try_from(value: TextureLayout) -> Result<Self, Self::Error> {
+        match value {
+                TextureLayout {
+                    top: None,
+                    bottom: None,
+                    ..
+                } => return Err(anyhow!("Each entry in texture_layouts in transformation \"fixed_position_layout\" requires either bottom or top coordinate.")),
+                TextureLayout {
+                    top: Some(_),
+                    bottom: Some(_),
+                    ..
+                } => return Err(anyhow!("Fields \"top\" and \"bottom\" are mutually exclusive, you can only specify one in texture layout in \"fixed_position_layout\" transformation.")),
+                _ => (),
+            };
+        match value {
+                TextureLayout {
+                    left: None,
+                    right: None,
+                    ..
+                } => return Err(anyhow!("Each entry in texture_layouts in transformation \"fixed_position_layout\" requires either right or left coordinate.")),
+                TextureLayout {
+                    left: Some(_),
+                    right: Some(_),
+                    ..
+                } => return Err(anyhow!("Fields \"left\" and \"right\" are mutually exclusive, you can only specify one in texture layout in \"fixed_position_layout\" transformation.")),
+                _ => (),
+            };
+
+        Ok(Self {
+            top: value.top.map_or(Ok(None), |v| v.try_into().map(Some))?,
+            bottom: value.bottom.map_or(Ok(None), |v| v.try_into().map(Some))?,
+            left: value.left.map_or(Ok(None), |v| v.try_into().map(Some))?,
+            right: value.right.map_or(Ok(None), |v| v.try_into().map(Some))?,
+            scale: value.scale.unwrap_or(1.0),
+            rotation: value.rotation.unwrap_or(Degree(0.0)).into(),
+        })
+    }
+}
+
+impl TryFrom<TiledLayout> for BuiltinSpec {
+    type Error = anyhow::Error;
+
+    fn try_from(layout: TiledLayout) -> Result<Self, Self::Error> {
+        let result = Self::TiledLayout(TiledLayoutSpec {
+            resolution: layout.resolution.into(),
+            background_color_rgba: layout
+                .background_color_rgba
+                .map(TryInto::try_into)
+                .unwrap_or(Ok(colors::RGBAColor(0, 0, 0, 0)))?,
+            tile_aspect_ratio: layout.tile_aspect_ratio.unwrap_or((16, 9)),
+            margin: layout.margin.unwrap_or(0),
+            padding: layout.padding.unwrap_or(0),
+            horizontal_alignment: layout
+                .horizontal_alignment
+                .unwrap_or(HorizontalAlign::Center)
+                .into(),
+            vertical_alignment: layout
+                .vertical_alignment
+                .unwrap_or(VerticalAlign::Center)
+                .into(),
+        });
```

Comment:
I think serde defaults are more readable than manual default implementations on parsing. Using serde, default values are already visible in the type definition, but with this approach, it's unclear what they are looking only at types.

Reply:
I would agree if there was a better way to do that more universally, but if we are using literals in some places, functions in other and in some cases it will need to be done inside `from()` function then I prefer to have that consistent at least.

>  Using serde, default values are already visible in the type definition

No, they are next to type definition (and if you are using functions that return default not even that), but they are not part of type definition. 

Also all those serde annotations are really affecting readability and size. It's especially important that top level API like e.g. entire Node structure would be visible without jumping between bunch of files. e.g. currently to see how to create node e.g. fixed_position_layout you need to check at least 4 files in compositor_common package to understand it's structure.

Reply:
There is alternative approach I considered that would work with your suggestion, but I'm still slightly leaning towards current approach.

What do you think about https://github.com/membraneframework/video_compositor/pull/194/files#r1360454689 ?

Reply:
I get your point. It's hard for me to tell whether wrapping everything in types and implementing defaults on them is better than implementing it in `from` functions. I leave it up to you.

Reply:
I think the current approach is better

Reply:
> I get your point. It's hard for me to tell whether wrapping everything in types and implementing defaults on them is better than implementing it in from functions. I leave it up to you.

That is not what I suggested, an alternative could be using `mod` blocks that wrap multiple types and using serde default mechanism not rust  Default trait

Reply:
> I think wrapping in types is slightly better since we still have the possibility to automatically generate types/schema for API.

Not sure what you mean by that

Reply:
Why can't we just use `struct FieldName(type)` and `impl Default for FieldName` on the API level? 
We could avoid implementing `try_into` everywhere (only when the field default value depends on other fields values we can't do that). 
I think some tools could automatically generate schema with default values for us.

Reply:
> Why can't we just use struct FieldName(type) and impl Default for FieldName on the API level?

We can, but why do you think it's better?

> We could avoid implementing try_into everywhere (only when the field default value depends on other fields values we can't do that).

This would not help with try_into, defaults are not related in any way to when we use From vs TryFrom trait

> I think some tools could automatically generate schema with default values for us.

Did you see my comment linked above https://github.com/membraneframework/video_compositor/pull/194/files#r1360454689 ?